### DISCUSSION TITLE
redux actions not working when dispatched inside of useEffects in dynamic routes / routes with query params

### DISCUSSION THREAD

**Discussion by Dizigen** [vercel/next.js#24946] [Help]
**What version of Next.js are you using?**

10.0.5

**What version of Node.js are you using?**

12.16.1

**What browser are you using?**

Chrome

**What operating system are you using?**

macOS

**How are you deploying your application?**

local development

**Describe the Bug**

When a redux action is dispatched inside of a useEffect hook, and the current page is a dynamic route, or a route with query parameters set, the redux states are not updated.

**Expected Behavior**

redux actions should mutate store values the same way as in static routes with no query params set.

**To Reproduce**

1. clone this repo: https://github.com/Dizigen/nextjs-redux-bug
2. run `yarn && yarn dev`
3. navigate to http://localhost:3000/test -> notice the email redux state is set properly
4. navigate to http://localhost:3000/test?foo=bar -> email redux state is not set (notice the expected email sometimes flashes for a brief moment before disappearing).
5.  navigate to a dynamic route http://localhost:3000/hello -> same as 4.

**Comment by kikoanis** [comment]
For me this is expected. The `App` will render twice due to the change in route. One for the root and one for the query.
This could be avoided if you add an `getInitialProps` or similar methods (static or server) to the App function. Something similar to this
```
App.getInitialProps= async (appContext) => {
  const { Component, ctx } = appContext;
  let pageProps = {};
  if (Component.getInitialProps) {
     pageProps = await Component.getInitialProps(ctx);
  }

  return { pageProps };
}
```

**Comment by sikhote** [comment]
I had this issue when trying to do a `router.replace` within a `useEffect`. Doing this caused the store to just ignore some actions from having an effect on the store. Hopefully this gets documented more, especially by redux.

**Comment by sikhote** [comment]
The issue I am having...

I run this code one page:

```
dispatch(actionsVerify.save({ status, addressData }));
Router.replace({
  pathname: '/verify',
  query: qs.parse(window.location.search),
});
```

The action is nothing fancy, just a regular action, no thunk usage, but I do use thunk middleware.

Then in the next page I try to access that data. About 20% of the time the data is not there. Any idea what is causing this?
